### DISCUSSION TITLE
How to integrate mongoose into nuxt

### DISCUSSION THREAD

**Discussion by mattiabonardi** [nuxt/nuxt#16987] [Questions]
Hello everybody,
I would like to use mongoose to get data from my database. 
I only found a dirty method which consists in using mongoose functions directly in the `server/api/index.ts` file but it doesn't work in production.
If I put a `server/index.ts` it doesn't run when server started. 

it is interesting to have a function available which is executed when the server starts where to put all the confguration including the connection opening with mongoose.

For example: `server/index.ts`
```typescript
mongoose
  .connect(
    "mongodb+srv://xxxxxxxxxxxxxxxxxxxxxxxx"
  )
  .then(() => {
    console.log("Connection to mongo estabilished");
  })
  .catch((error) => {
    console.error("Error during mongo connection", error);
  });
```

Thank you

**Comment by Diizzayy** [comment]
@mattia-bonardi-smeup You can create a directory under server called `db` to house your database setup / functions. 

Below is a working example that I knocked up

`nuxt.config.ts`
```ts
import { defineNuxtConfig } from "nuxt3";

export default defineNuxtConfig({
  publicRuntimeConfig: {
    MONGO_URI: "<DATABASE_URI>"
  },
});
```

`~/server/db/index.ts`
```ts
import config from "#config";
import mongoose from "mongoose";

mongoose.connect(config.MONGO_URI, () => {
  console.log("Connected to MongoDB");
});

export type IPost = {
  _id?: string;
  image?: string;
  title: string;
  message: string;
  author?: string;
  status: "draft" | "published";
  createdAt?: string;
  updatedAt?: string;
}

const PostSchema = new mongoose.Schema<IPost>(
  {
    image: { type: String },
    title: { type: String, required: true },
    message: { type: String, required: true },
    author: String,
    status: {
      type: String,
      enum: ["draft", "published"],
      default: "published",
    },
  },
  { timestamps: true }
);

export type PostQuery = mongoose.FilterQuery<IPost>;
export const PostModel = mongoose.model<IPost>("Post", PostSchema);
```

`~/server/api/articles`
```ts
import { PostModel } from "../db";
import type { IPost, PostQuery } from "../db";

export default async (req, res): Promise<IPost[]> => {
  let query: PostQuery = { status: "published" };
  return await PostModel.find(query);
};
```

@mattia-bonardi-smeup This example works exactly as you'd expect.

**Comment by mattiabonardi** [comment]
Thank you @Diizzayy for the answer. 
I tried it and it works fine with `nuxi dev` command but when I builds and starts the application it doesen't work. 

It doesn't work also when put into `~/server/db/index.ts` only `mongoose.connect()` method.

**Reply by Diizzayy** [reply]
Strange. It works and builds properly in production mode for me

**Reply by mattiabonardi** [reply]
In your case mongo connection is estabilished when the application starts or when you call `/api/articles` api ?

**Reply by Diizzayy** [reply]
Indeed, and It works perfectly in both dev and production mode

**Comment by john-prutton** [comment]
For an updated solution, see nuxt/framework#5525.
